File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.java
class Solution {

    private List<List<Integer>> res;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        //start by sorting everything
        Arrays.sort(candidates);
        // then we proceed to the same backtracking algorithm, it's always the dfs tree,
        // initialize all of the arrays I will need
        res = new ArrayList<>();
        backtrack(0, new ArrayList<>(), target, 0, candidates);
        return res;
    }

    public void backtrack(int i, List<Integer> sub, int target, int total, int[] candidates){
        
        if(target == total){
            res.add(new ArrayList<>(sub));
            return;
        }

        if(target < total || i >= candidates.length){
            return;
        }

        sub.add(candidates[i]);
        backtrack(i+1, sub, target, total+candidates[i], candidates);
        sub.remove(sub.size()-1);

        while(i < candidates.length -1 && candidates[i] == candidates[i+1] ){
            i++;
        }
        backtrack(i+1, sub, target, total, candidates);
    }


}

File: Data Structures & Algorithms/interleaving-string/submission-2.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False

        dp = {}

        def backtrack(i, i1, i2):
            if i == len(s3):
                return (i1 == len(s1) and i2 == len(s2))
            
            if (i1,i2) in dp:
                return dp[(i1,i2)]

            res = False
            if i1 < len(s1) and s1[i1] == s3[i]:
                res = backtrack(i+1, i1+1, i2)

            if not res and i2 < len(s2) and s2[i2]==s3[i]:
                res = backtrack(i+1, i1, i2+1)

            dp[(i1,i2)] = res
            return res
 

        return backtrack(0,0,0)